### PR TITLE
[api-extractor] correcting edge case on tokenRanges

### PR DESCRIPTION
**Pull request by mrshllstock**
## Summary

In #3317 I addressed an error for tokenRanges from having the last token come from a separator at the end of recursion and when updating our library version came across this edge case.

This leads to two summary points:
1) We should ensure that we don't have a situation where excerptStartIndex can equal excerptEndIndex, as the indexes are  [excerptStartIndex, excerptEndIndex), hence the logic for the fix.
2) Replicating the issue in a rushStack api-extractor-scenario with a class such as this
```
/** This is my class. */
class MyClass {
   * This method does something.
   *
   * @param {[number]} This is x.
   * @param {[string]} This is y.
   * @return {[boolean]} Returns a result.
   */
  someMethod(x , y )  {
    return false;
  }
}
```

Is leading to a returnTypeTokenRange of [4,5), where internally we are seeing [5,5). Is there a parameter that changes the handling of extraneous spaces?

### CODE REVIEW

File: apps/api-extractor/src/generators/ExcerptBuilder.ts
Comment (octogonz) [inline apps/api-extractor/src/generators/ExcerptBuilder.ts:204]: > Is leading to a returnTypeTokenRange of [4,5), where internally we are seeing [5,5). 
> Is there a parameter that changes the handling of extraneous spaces?

I'm not sure what you're asking about.  What kind of parameter do you mean?

Comment (octogonz) [reply apps/api-extractor/src/generators/ExcerptBuilder.ts:204]: > Replicating the issue in a rushStack api-extractor-scenario with a class such as this

If you could add this test case to your PR branch, that would be super helpful. :-)

File: apps/api-extractor/src/generators/ExcerptBuilder.ts
Comment (mrshllstock) [reply apps/api-extractor/src/generators/ExcerptBuilder.ts:201]: I can change this to the preferable expression, and yes to my understanding `excerptEndIndex` should always be larger than `excerptStartIndex`

File: apps/api-extractor/src/generators/ExcerptBuilder.ts
Comment (mrshllstock) [reply apps/api-extractor/src/generators/ExcerptBuilder.ts:204]: Will do once I understand the reasoning for the difference in output; as atm without any dts rollup options (*trimmedFilePath from https://api-extractor.com/pages/setup/configure_rollup/) declared we are seeing extra spaces in the rollup.d.ts which is leading to the edge case we are seeing. For the code segment above we would end up with the d.ts looking like.
```
someMethod(x : number , y : string ) : boolean ;
```
where as API extractor is expecting the method signature to look more like
```
someMethod(x: number, y: string): boolean;
```

Comment (mrshllstock) [reply apps/api-extractor/src/generators/ExcerptBuilder.ts:204]: Alright more details as this developed and the like.

I added a new api-extractor-test that runs on a predetermined d.ts with additional spaces, which is valid typescript, just different than the output of TSC.

I hope this illuminates the edge case appropriately and is ready for further review.

File: build-tests/api-extractor-test-05/lib/index.d.ts
Comment (zelliott) [inline build-tests/api-extractor-test-05/lib/index.d.ts:20]: Some of these names are a little odd (e.g. `testJsLibrary` and `testJsLibrary_MyClass`) - I know they're taken from our specific internal instance of this bug but maybe we can rename them something more generic here? Also, consider simplifying this test down to a truly minimal test case to make it clearer what we're exactly testing (maybe just an exported variable and exported function with extra spaces on both).

File: build-tests/api-extractor-test-05/etc/api-extractor-test-05.api.md
Comment (zelliott) [inline build-tests/api-extractor-test-05/etc/api-extractor-test-05.api.md:10]: This looks odd... is this correct?

File: build-tests/api-extractor-test-05/etc/api-extractor-test-05.api.json
Comment (zelliott) [inline build-tests/api-extractor-test-05/etc/api-extractor-test-05.api.json:231]: I think in an ideal world, we'd want this:

```
{
    "kind": "Content",
    "text": "someMethod(x : "
},
{
    "kind": "Content",
    "text": "number "
},
```

to really be this:

```
{
    "kind": "Content",
    "text": "someMethod(x : "
},
{
    "kind": "Content",
    "text": "number"
},
{
    "kind: "Content",
    "text": " "
}
```

That way it's possible to construct a param type range that only grabs `number` and not `number `. However, my guess is that this may be non-trivial work and potentially also not really worth it at this time if the tool producing the `.d.ts` is already adding _extraneous_ spaces. So, feel free to ack & close, but just thought I'd mention this.

File: apps/api-extractor/build-tests.cmd
Comment (zelliott) [inline apps/api-extractor/build-tests.cmd:3]: Consider naming this test something more informative like `api-extractor-extra-spaces-dts` (happy to bikeshed) instead of `api-extractor-test-05`. That being said, I know there's precedent for the naming you chose given the current test names, so I don't feel strongly (maybe @octogonz does).

File: build-tests/api-extractor-test-05/config/api-extractor.json
Comment (zelliott) [inline build-tests/api-extractor-test-05/config/api-extractor.json:11]: Where is the dts rollup? Is it purposefully not added as `dist/` is ignored by Git or something?

**Review by zelliott — APPROVED** [review]
LGTM

File: build-tests/api-extractor-test-05/etc/api-extractor-test-05.api.md
Comment (mrshllstock) [reply build-tests/api-extractor-test-05/etc/api-extractor-test-05.api.md:10]: This does look odd, but I imagine this is since the variable is deprecated and let's are capable of being multi-line this is how the markdown handles included the deprecated handler inline

File: build-tests/api-extractor-test-05/lib/index.d.ts
Comment (mrshllstock) [reply build-tests/api-extractor-test-05/lib/index.d.ts:20]: I left in just enough to have multiple instances in the commits showing the error state for further verbosity, and figured that the additional 'bloat' would service to be more akin to a real scenario.

Minimizing this now.

File: build-tests/api-extractor-test-05/config/api-extractor.json
Comment (mrshllstock) [reply build-tests/api-extractor-test-05/config/api-extractor.json:11]: Yeah the git is structured to ignore the `dist` directory; however it is 1:1 (spaces included) of our index.d.ts with an additional `export { }` appended to the end

File: apps/api-extractor/build-tests.cmd
Comment (mrshllstock) [reply apps/api-extractor/build-tests.cmd:3]: I left it to follow the nomenclature in place, but I can definitely rename it to be more semantic.

File: build-tests/api-documenter-test/etc/api-documenter-test.api.json
Comment (zelliott) [inline build-tests/api-documenter-test/etc/api-documenter-test.api.json:170]: What is this? I can't find it anywhere in the repo.

File: build-tests/api-extractor-test-05/etc/api-extractor-test-05.api.json
Comment (mrshllstock) [reply build-tests/api-extractor-test-05/etc/api-extractor-test-05.api.json]: I'm not sure offhand if there'd be a clean way to handle that, unless we want to only lead to teh additional spaces for the last of a token range to add the space to the next, since I could see the argument for `someMethod(x : ` to have teh last space moved to the next token.

File: build-tests/api-documenter-test/etc/api-documenter-test.api.json
Comment (mrshllstock) [reply build-tests/api-documenter-test/etc/api-documenter-test.api.json:170]: Good Q: this started being populated once I did the merge yesterday to resolve a conflict, so I was under the impression this was a recently included feature.

Comment (zelliott) [reply build-tests/api-documenter-test/etc/api-documenter-test.api.json:170]: @octogonz Ideas? @mrshllstock, if you rebase your branch off the latest changes from main, does this go away?

File: apps/api-extractor/build-tests.cmd
Comment (zelliott) [reply apps/api-extractor/build-tests.cmd:3]: Gotcha, I'll leave it up to you.

**Comment by zelliott** [comment]
It looks like the CI Build is failing, but perhaps the build details have since expired because it was last run a few weeks ago.

File: build-tests/api-documenter-test/etc/api-documenter-test.api.json
Comment (mrshllstock) [reply build-tests/api-documenter-test/etc/api-documenter-test.api.json]: I rebased main (boy was that an ugly PR state for a moment). We can see this has been added in other tests from api-documenter, such as the tsdocConfig test.

Comment (zelliott) [reply build-tests/api-documenter-test/etc/api-documenter-test.api.json]: Ah I see, this was added in the TSDoc repo (https://github.com/microsoft/tsdoc/pull/317).

File: build-tests/api-extractor-test-05/typings/tsd.d.ts
Comment (zelliott) [inline build-tests/api-extractor-test-05/typings/tsd.d.ts:7]: Is this file needed?

**Comment by zelliott** [comment]
Still LGTM. A few open threads still on this PR that it'd be nice to get @octogonz's feedback on, otherwise I think this is close to going in.

File: build-tests/api-extractor-scenarios/src/excerptTokens/index.d.ts
Comment (octogonz) [inline build-tests/api-extractor-scenarios/src/excerptTokens/index.d.ts:1]: I was able to eliminate the `api-extractor-test-05`  project by using `// prettier-ignore` in the .d.ts file and adding a step to `runScenarios.ts` that copies the .d.ts files to `lib/`

@zelliott

**Comment by octogonz** [comment]
@mrshllstock thanks for contributing this fix!  👍👍

**Review by octogonz — APPROVED** [review]

**Comment by mrshllstock** [comment]
> @mrshllstock thanks for contributing this fix! 👍👍

Thank you for making it cleaner and more consistent with the structure of scenarios

File: build-tests/api-extractor-test-05/etc/api-extractor-test-05.api.json
Comment (octogonz) [reply build-tests/api-extractor-test-05/etc/api-extractor-test-05.api.json]: I'm most concerned that the normalized compiler output is handled correctly, since this is our primary use case. If people want to do additional work to normalize the output, that's cool.

HOWEVER there was some discussion about running a preprocessor such as Prettier on the .d.ts content (before and/or after) API Extractor processes them.  That could be a "battle axe" that gets us out of having to deal with a lot of these edge cases.  (Code formatting is a lot trickier than it seems!)

**Comment by octogonz** [comment]
Released with API Extractor 7.24.0

**Pull request merged**
